### ISSUE TITLE
Maybe a bug or not! Spicedb connect to Pgbouncer report error "prepared statement name is already in use (SQLSTATE 08P01)"

### ISSUE THREAD

**Issue by LucasZhanye** [authzed/spicedb#1992]
Recently, I encountered a problem，I have a postgresql cluster which build by pigsty, and I use pgbouncer as pool.But when I use docker to run spicedb, command is "docker run --name spicedb2 --network host authzed/spicedb:v1.33.0-amd64 datastore migrate head --datastore-engine=postgres --datastore-conn-uri="postgres://zzy_test:dsfsgxawrt@192.168.120.110:5433/zzy_test",I always receive an error "prepared statement name is already in use (SQLSTATE 08P01)"

And I have create a issue for Pgbouncer,this is the link: https://github.com/pgbouncer/pgbouncer/issues/1118

The issue mentions that the application(spicedb)  use an already used prepared statement name, they suggest me to fix the application to not repeat the same prepared statement name for the same client. 

Because of I use docker container for spicedb, I do not know how to solve this error.

**Comment by vroldanbet** [comment]
Hey @LucasZhanye, thanks for reporting.

SpiceDB does not support pgbouncer out of the box without some tweaking. Some of this has been captured in https://github.com/authzed/spicedb/issues/1217.

SpiceDB mas use of `pgx` client side option `default_query_exec_mode=cache_statement`, which reads as follows:
>	Automatically prepare and cache statements. This uses the extended protocol. Queries are executed in a single round
	trip after the statement is cached. This is the default. If the database schema is modified or the search_path is
	changed after a statement is cached then the first execution of a previously cached query may fail. e.g. If the
	number of columns returned by a "SELECT *" changes or the type of a column is changed.
	
The documentation does not indicate it explicitly in this mode of operation, but does document in other modes which also lead to 2 rountrips that:

>This may cause problems with connection poolers that switch the underlying connection between round trips

Effectively SpiceDB cannot operate without the guarantee that the same underlying connection is used. The fact you got a prepared statement collision could stem from this.

Some users reported in https://github.com/authzed/spicedb/issues/1217 that configuring pgbouncer with [`pool_mode = session`](https://www.pgbouncer.org/config.html#pool_mode) would address this. I haven't tested it myself, but from the documentation it reads like the behaviour needed to avoid switching underlying connections. The `ignore_startup_parameters=plan_cache_mode` also must be set, since it's not supported, which would also impact query plans in its current form, so you are likely to get some suboptimal query plans.

A potential long-term solution is to stop using prepared statements altogether, given they provide no benefit to SpiceDB because of the variability of the queries generated by it. I happen to be working on this via https://github.com/authzed/spicedb/pull/1989, where I described the motives for the change.

So in short, give those parameters a go, but right now using pgbouncer is not recommended until support for connection pooler-friendly execution mode (which is on the works via https://github.com/authzed/spicedb/pull/1989) is introduced. You are likely get subpar performance via pgbouncer right now.

I'm going to close this as a dupe of https://github.com/authzed/spicedb/issues/1217

**Comment by LucasZhanye** [comment]
> Duplicate of #1217
> 
> Hey @LucasZhanye, thanks for reporting.
> 
> SpiceDB does not support pgbouncer out of the box without some tweaking. Some of this has been captured in #1217.
> 
> SpiceDB mas use of `pgx` client side option `default_query_exec_mode=cache_statement`, which reads as follows:
> 
> > Automatically prepare and cache statements. This uses the extended protocol. Queries are executed in a single round
> > trip after the statement is cached. This is the default. If the database schema is modified or the search_path is
> > changed after a statement is cached then the first execution of a previously cached query may fail. e.g. If the
> > number of columns returned by a "SELECT *" changes or the type of a column is changed.
> 
> The documentation does not indicate it explicitly in this mode of operation, but does document in other modes which also lead to 2 rountrips that:
> 
> > This may cause problems with connection poolers that switch the underlying connection between round trips
> 
> Effectively SpiceDB cannot operate without the guarantee that the same underlying connection is used. The fact you got a prepared statement collision could stem from this.
> 
> Some users reported in #1217 that configuring pgbouncer with [`pool_mode = session`](https://www.pgbouncer.org/config.html#pool_mode) would address this. I haven't tested it myself, but from the documentation it reads like the behaviour needed to avoid switching underlying connections. The `ignore_startup_parameters=plan_cache_mode` also must be set, since it's not supported, which would also impact query plans in its current form, so you are likely to get some suboptimal query plans.
> 
> A potential long-term solution is to stop using prepared statements altogether, given they provide no benefit to SpiceDB because of the variability of the queries generated by it. I happen to be working on this via #1989, where I described the motives for the change.
> 
> So in short, give those parameters a go, but right now using pgbouncer is not recommended until support for connection pooler-friendly execution mode (which is on the works via #1989) is introduced. You are likely get subpar performance via pgbouncer right now.
> 
> I'm going to close this as a dupe of #1217

OK，Thank you for reply!
